File: src/clj_peg/peg_in_peg.clj
(ns clj-peg.peg-in-peg
  (:use clj-peg.peg)
  (:use clj-peg.grammar)
  (:use clj-peg.derived-rules))

;; I don't really like this because it uses eval, but I don't know how to
;; dynamically let something any other way
(defn build-let-fn [p]
  (fn [b]
    (eval
     `(let [~@(apply concat b)]
	~p))))

(defrule rule-def
  (mkret
   (mkseq
    [(mkbind (mkrule 'rule-name) 'rule-name)
     (mklit '=)
     (mkbind (mkrule 'alternatives) 'rule-body)
     (mkrule 'end)])
   (fn [b]
     [{:rule-name (first (b 'rule-name))
       :rule-body (first (b 'rule-body))}])))

; a symbol but not a keyword
(defrule rule-name
  (mkseq
   [(mknot (mkrule 'peg-keyword))
    (mkrule 'match-symbol)]))

(defrule sequence-match
     (mkret
      (mkseq
       [(mkseq
	 (mk1om (mkrule 'rule-element))
	 'rules)
	(make-opt-matcher
	 (make-sequence-matcher
	  [(make-rule-matcher 'clj-peg.peg-in-peg/literal->)
	   (make-bind
	    (make-rule-matcher 'clj-peg.peg-in-peg/return-expression)
	    'r-exp)]))])
      (fn [b]
	(let [rules (b 'rules)
	      exp (first (b 'r-exp))]
	  [(if exp
	     (make-return
	      (make-sequence-matcher rules)
	      exp)
	     (make-sequence-matcher rules))]))))

(def return-expression
     (make-return
      (make-rule-matcher 'clj-peg.derived-rules/anything)
      (fn [b]
	(let [exp (first (b '-match-))]
	  [(build-let-fn exp)]))))

(def alternatives
     (make-return
      (make-sequence-matcher
       [(make-zero-or-more-matcher
	 (make-sequence-matcher
	  [(make-rule-matcher 'clj-peg.peg-in-peg/sequence-match)
	   (make-rule-matcher 'clj-peg.peg-in-peg/literal-slash)]))
	(make-rule-matcher 'clj-peg.peg-in-peg/sequence-match)])
      (fn [b]
	(let [alts (b '-match-)]
	  [(make-alt-matcher
	    (filter #(not (= '/ %)) alts))]))))

(def rule-element
     (make-alt-matcher
      [(make-rule-matcher 'clj-peg.peg-in-peg/binding-match)
       (make-rule-matcher 'clj-peg.peg-in-peg/optional)
       (make-rule-matcher 'clj-peg.peg-in-peg/zero-or-more)
       (make-rule-matcher 'clj-peg.peg-in-peg/one-or-more)
       (make-rule-matcher 'clj-peg.peg-in-peg/rule-element-atomic)]))

(def binding-match
     (make-return
      (make-sequence-matcher
       [(make-bind (make-rule-matcher 'clj-peg.peg-in-peg/rule-element-atomic) 'rule)
	(make-rule-matcher 'clj-peg.peg-in-peg/literal=>)
	(make-bind (make-rule-matcher 'clj-peg.peg-in-peg/variable) 'var)])
      (fn [b]
	(let [var (first (b 'var))
	      rule (first (b 'rule))]
	  [(make-bind rule var)]))))

(def optional
     (make-return
      (make-sequence-matcher
       [(make-bind (make-rule-matcher  'clj-peg.peg-in-peg/rule-element-atomic) 'rule)
	(make-rule-matcher 'clj-peg.peg-in-peg/literal?)])
      (fn [b]
	(let [rule (first (b 'rule))]
	  [(make-opt-matcher rule)]))))

(def zero-or-more
     (make-return
      (make-sequence-matcher
       [(make-bind (make-rule-matcher 'clj-peg.peg-in-peg/rule-element-atomic) 'rule)
	(make-rule-matcher 'clj-peg.peg-in-peg/literal*)])
      (fn [b]
	(let [rule (first (b 'rule))]
	  [(make-zero-or-more-matcher rule)]))))

(def not-rule
     (make-return
      (make-sequence-matcher
       [(make-rule-matcher  'clj-peg.peg-in-peg/literal--)
	(make-bind (make-rule-matcher  'clj-peg.peg-in-peg/rule-element-atomic) 'rule)])
      (fn [b]
	(let [rule (first (b 'rule))]
	  [(make-not rule)]))))

(def one-or-more
     (make-return
      (make-sequence-matcher
       [(make-bind (make-rule-matcher 'clj-peg.peg-in-peg/rule-element-atomic) 'rule)
	(make-rule-matcher  'clj-peg.peg-in-peg/literal+)])
      (fn [b]
	(let [rule (first (b 'rule))]
	  [(make-one-or-more-matcher rule)]))))

(def rule-element-atomic
     (make-alt-matcher
      [(make-rule-matcher 'clj-peg.peg-in-peg/predicate)
       (make-rule-matcher 'clj-peg.peg-in-peg/parenthesis)
       (make-rule-matcher 'clj-peg.peg-in-peg/rule-match)
       (make-rule-matcher 'clj-peg.peg-in-peg/string-matcher)
       (make-rule-matcher 'clj-peg.peg-in-peg/atom-match)
       (make-rule-matcher 'clj-peg.peg-in-peg/not-rule)]))

(def atom-match
     (make-return
      (make-alt-matcher
       [(make-rule-matcher 'clj-peg.derived-rules/match-number)
	(make-rule-matcher 'clj-peg.derived-rules/match-char)])
      (fn [b]
	(let [a (first (b 'a))]
	  [(make-literal-matcher a)]))))

(def predicate
     (make-return
      (make-rule-matcher 'clj-peg.derived-rules/match-list)
      (fn [b]
	(let [p (first (b '-match-))]
	  [(make-pred-matcher (build-let-fn p))]))))

(def parenthesis
     (make-return
      (make-rule-matcher 'clj-peg.derived-rules/match-vec)
      (fn [b]
	(let [v (first (b '-match-))
	      r (alternatives v b)]
	  (:r r)))))

(def rule-match
     (make-return
      (make-sequence-matcher
       [(make-not (make-rule-matcher 'clj-peg.peg-in-peg/peg-keyword))
	(make-rule-matcher 'clj-peg.derived-rules/match-symbol)])
      (fn [b]
	(let [s (first (b '-match-))]
	  [(make-rule-matcher  s)]))))

(def string-matcher
     (make-return
      (make-rule-matcher 'clj-peg.derived-rules/match-string)
      (fn [b]
	(let [s (first (b '-match-))]
	  [(make-string-matcher s)]))))

(def variable
     (make-sequence-matcher
      [(make-not (make-rule-matcher 'clj-peg.peg-in-peg/peg-keyword))
       (make-rule-matcher 'clj-peg.derived-rules/match-symbol)]))

(def peg-keyword
     (make-alt-matcher
      [(make-rule-matcher 'clj-peg.peg-in-peg/literal=)
       (make-rule-matcher 'clj-peg.peg-in-peg/literal->)
       (make-rule-matcher 'clj-peg.peg-in-peg/literal?)
       (make-rule-matcher 'clj-peg.peg-in-peg/literal*)
       (make-rule-matcher 'clj-peg.peg-in-peg/literal+)
       (make-rule-matcher 'clj-peg.peg-in-peg/literal=>)
       (make-rule-matcher 'clj-peg.peg-in-peg/literal--)
       (make-rule-matcher 'clj-peg.peg-in-peg/literal-slash)]))

(def literal= (make-literal-matcher '=))
(def literal-> (make-literal-matcher '->))
(def literal? (make-literal-matcher '?))
(def literal* (make-literal-matcher '*))
(def literal+ (make-literal-matcher '+))
(def literal=> (make-literal-matcher '=>))
(def literal-- (make-literal-matcher '--))
(def literal-slash (make-literal-matcher '/))

;; a macro to create a new rule
  
(defmacro dr [& d]
  (let [r  (rule-def d {})]
    (if (nil? r)
      (throw (RuntimeException. "not a valid grammar"))
      (let [rule (first (:r r))]
       `(def ~(:rule-name rule) ~(:rule-body rule))))))

File: test/clj_peg/peg_in_peg.c
(ns clj-peg.peg_in_peg_test
  (:use [clj-peg.peg])
  (:use [clj-peg.grammar])
  (:use [clj-peg.derived-rules])
  (:use [clj-peg.peg-in-peg] :reload-all)
  (:use [clojure.test]))

(deftest test-keyword-literals
  (dorun
   (map #(is ((lookup *peg* %) [%] {}))
	keyword-list)))

(deftest test-keyword-rule
  (dorun
   (map #(is ((lookup *peg* 'keyword) [%] {}))
	keyword-list)))

(deftest test-variable
  (is (= ['hh] (:r ((lookup *peg* 'variable) ['hh] {}))))
  (is (not ((lookup *peg* 'variable) [] {})))
  (is (not ((lookup *peg* 'variable) [0] {}))))

(deftest test-string
  (is (= ["abc"] (:r ((lookup *peg* 'string) ["abc"] {}))))
  (is (not ((lookup *peg* 'string) [5] {}))))

(deftest test-symbol
  (is (= ['hh] (:r ((lookup *peg* 'symbol) ['hh] {}))))
  (is (not ((lookup *peg* 'symbol) [] {})))
  (is (not ((lookup *peg* 'symbol) [0] {}))))

(deftest test-string-match
  (is (not ((lookup *peg* 'string-match) [1] {})))
  (is (ifn? (first (:r ((lookup *peg* 'string-match) ["abc"] {})))))
  (is (= (vec "abc") (:r ((first (:r ((lookup *peg* 'string-match) ["abc"] {}))) "abc" {}))))
  (is (not ((first (:r ((lookup *peg* 'string-match) ["abc"] {}))) "acb" {}))))

(deftest test-rule-match
  (is (ifn? (first (:r ((lookup *peg* 'rule-match) ['anything] {})))))
  (is (not ((lookup *peg* 'rule-match) [0 3] {})))
  (is (= [1] (:r ((first (:r ((lookup *peg* 'rule-match) ['anything] {}))) [1] {}))))
  (is (not ((first (:r ((lookup *peg* 'rule-match) ['end] {}))) [1] {}))))

(deftest test-list
  (is (not ((lookup *peg* 'list) [1] {})))
  (is (= ['(1 2 3)] (:r ((lookup *peg* 'list) ['(1 2 3)] {})))))

(deftest test-vec
  (is (not ((lookup *peg* 'vec) [1] {})))
  (is (= [[1 2 3]] (:r ((lookup *peg* 'vec) [[1 2 3]] {})))))

(deftest test-parenthesis
  (is (not ((lookup *peg* 'parenthesis) [1] {})))
  (is (= [1 2 3] (:r ((first (:r ((lookup *peg* 'parenthesis) [['anything 'anything 'anything]] {}))) [1 2 3] {}))))
  (is (= [3] ((:b ((first (:r ((lookup *peg* 'parenthesis) '([anything anything anything => b]) {}))) [1 2 3] {})) 'b))))

(deftest test-predicate
  (is (not ((lookup *peg* 'predicate) [1] {})))
  (is (ifn? (first (:r ((lookup *peg* 'predicate) ['(even? p)] {})))))
  (is ((first (:r ((lookup *peg* 'predicate) ['(= 1 1)] {}))) [] {}))
  (is (not ((first (:r ((lookup *peg* 'predicate) ['(= 1 2)] {}))) [] {}))))

(deftest test-atom
  (is ((lookup *peg* 'atom) [1] {}))
  (is ((lookup *peg* 'atom) "a" {})))

(deftest test-rule-element-atomic
  (dorun
   (map #(is ((lookup *peg* 'rule-element-atomic) [%] {}))
	'((even? a b)
	  [1 2 3]
	    anything
	    "abc"
	    1
	    \a
	    )))
  (is ((lookup *peg* 'rule-element-atomic) '(-- anything) {})))

(deftest test-1-o-m
  (is ((lookup *peg* 'one-or-more) '(anything +) {}))
  (is (not ((lookup *peg* 'one-or-more) '(anything) {})))
  (is (not ((lookup *peg* 'one-or-more) '(anything anything +) {})))
  (is (= [1 2 3] (:r ((first (:r ((lookup *peg* 'one-or-more) '(anything +) {}))) [1 2 3] {})))))

(deftest test-not
  (is ((lookup *peg* 'not) '(-- anything) {}))
  (is (not ((lookup *peg* 'not) '(anything) {})))
  (is ((first (:r ((lookup *peg* 'not) '(-- anything) {}))) [] {}))
  (is (not ((first (:r ((lookup *peg* 'not) '(-- anything) {}))) [1] {}))))

(deftest test-0-o-m
  (is ((lookup *peg* 'zero-or-more) '(anything *) {}))
  (is ((first (:r ((lookup *peg* 'zero-or-more) '(anything *) {}))) [] {}))
  (is ((first (:r ((lookup *peg* 'zero-or-more) '(anything *) {}))) [1] {}))
  (is ((first (:r ((lookup *peg* 'zero-or-more) '(anything *) {}))) [1 2] {})))
